### PR TITLE
Horizontal bar chart

### PR DESCRIPTION
**Pull request by deepaknevdepsl**
### Description
Added Horizontal Bar chart as a new visualization type.

### Issues Resolved
#885

### Check List
- [ ] New functionality includes testing.
  - [ ] All tests pass, including unit test, integration test and doctest
- [ ] Commits are signed per the DCO using --signoff

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: dashboards-observability/public/components/visualizations/charts/bar/bar.tsx [hunk 1]
```diff
@@ -26,15 +25,16 @@ export const Bar = ({ visualizations, layout, config }: any) => {
...
-    visualizations.data.rawVizData.bar.dataConfig;
+    visualizations.data?.rawVizData?.[visType]?.dataConfig &&
+    visualizations.data.rawVizData[visType].dataConfig;
   const xaxis = dataConfigTab?.dimensions
     ? dataConfigTab.dimensions.filter((item) => item.label)
     : [];
   const yaxis = dataConfigTab?.metrics ? dataConfigTab.metrics.filter((item) => item.label) : [];
   const barOrientation = dataConfig?.chartStyles?.orientation || vis.orientation;
-  const isVertical = barOrientation === vis.orientation;
+  let isVertical = visType === visChartTypes.HorizontalBar ? barOrientation !== vis.orientation : barOrientation === vis.orientation;
```
Comment (spattnaik) [inline dashboards-observability/public/components/visualizations/charts/bar/bar.tsx]: What is the requirement of changing `isVertical` from const to let?

File: dashboards-observability/public/components/visualizations/charts/helpers/viz_types.ts [hunk 2]
```diff
@@ -44,9 +44,14 @@ export const getVizContainerProps = ({
...
 }: IVizContainerProps): IVisualizationContainerProps => {
+  const isLineOrScatter = (vs: string) => (vs === visChartTypes.Line || vs === visChartTypes.Scatter);
+  const isVerticalorHorizontalBar = (vs: string) => (vs === visChartTypes.Bar || vs === visChartTypes.HorizontalBar);
+
   const getVisTypeData = () => {
-    if (vizId === visChartTypes.Line || vizId === visChartTypes.Scatter) {
+    if (isLineOrScatter(vizId)) {
       return vizId === visChartTypes.Line ? { ...getVisType(vizId, { type: visChartTypes.Line }) } : { ...getVisType(vizId, { type: visChartTypes.Scatter }) };
+    } else if (isVerticalorHorizontalBar(vizId)) {
+      return vizId === visChartTypes.Bar ? { ...getVisType(vizId, { type: visChartTypes.Bar }) } : { ...getVisType(vizId, { type: visChartTypes.HorizontalBar }) };
```
Comment (spattnaik) [inline dashboards-observability/public/components/visualizations/charts/helpers/viz_types.ts]: You are doing exact same steps in config_panel.tsx. Why not use this function there?

File: dashboards-observability/public/components/visualizations/charts/bar/bar_type.ts [hunk 3]
```diff
@@ -17,19 +17,19 @@ import {
...
 const { LegendPosition, ShowLegend } = DefaultChartStyles;
 export const createBarTypeDefinition = (params: any) => ({
-  name: 'bar',
+  name: params.type ? params.type : 'bar',
   type: 'bar',
-  id: 'bar',
-  label: 'Bar',
-  fullLabel: 'Bar',
+  id: params.type ? params.type : 'bar',
+  label: params.type === visChartTypes.HorizontalBar ? 'Horizontal Bar' : 'Bar',
```
Comment (spattnaik) [inline dashboards-observability/public/components/visualizations/charts/bar/bar_type.ts]: Don't we need to change the labels as `Vertical Bar`? What would be the difference between `Bar` and `Horizontal Bar`? `Horizontal Bar` is also a `Bar` anyway.

**Review by spattnaik — CHANGES_REQUESTED** [review]

File: dashboards-observability/public/components/visualizations/charts/helpers/viz_types.ts (hunk 2 above)
Comment (deepaknevdepsl) [reply dashboards-observability/public/components/visualizations/charts/helpers/viz_types.ts]: Created common function for this block of code.

File: dashboards-observability/public/components/visualizations/charts/bar/bar_type.ts (hunk 3 above)
Comment (deepaknevdepsl) [reply dashboards-observability/public/components/visualizations/charts/bar/bar_type.ts]: Yes, we have renamed Bar to Vertical Bar in next PR.

File: dashboards-observability/public/components/visualizations/charts/bar/bar.tsx (hunk 1 above)
Comment (deepaknevdepsl) [reply dashboards-observability/public/components/visualizations/charts/bar/bar.tsx]: Reverted as const, I was updating that value but after that I have changed that logic, and I forgot to revert this change.

File: dashboards-observability/public/components/event_analytics/explorer/visualizations/config_panel/config_panel.tsx [hunk 4]
```diff
@@ -222,20 +222,10 @@ export const ConfigPanel = ({ visualizations, setCurVisId, callback, changeIsVal
...
-        visDefinition = vs === visChartTypes.Line ? getVisType(vs, { type: visChartTypes.Line }) : getVisType(vs, { type: visChartTypes.Scatter });
-      } else if (isVerticalorHorizontalBar(vs)) {
-        visDefinition = vs === visChartTypes.Bar ? getVisType(vs, { type: visChartTypes.Bar }) : getVisType(vs, { type: visChartTypes.HorizontalBar });
-      } else {
-        visDefinition = getVisType(vs);
-      }
-
+      visDefinition = getVisTypeData(vs);
       return {
         ...visDefinition,
```
Comment (spattnaik) [inline dashboards-observability/public/components/event_analytics/explorer/visualizations/config_panel/config_panel.tsx]: I do not get it about the spread operator. Why do you have to use so much of it? Your function `getVisTypeData` calls some `getVisType` function and spreads the object to create a new object. Now you get the data into `visDefinition` and create another object by spreading that object. Aren't we overusing it?

Function purity is defined as - The function should not update any variables that are present outside the function or that are passed as variables. 
The variables it has created, what is the point in creating another variable to return it? What you are really doing here is 
```JS
visDefinition = getVisTypeData(vs);
const temp = {...visDefinition}; // you are just not seeing this variable, i.e. you had not explicitly defined it. But the computer has to do it.
return temp;
```
As your target function is returning a renewed object, you can simply return that. Or make your target function to return whatever it has received and create a copy when you update it.

File: dashboards-observability/public/components/event_analytics/explorer/visualizations/config_panel/config_panel.tsx [hunk 5]
```diff
@@ -222,20 +222,10 @@ export const ConfigPanel = ({ visualizations, setCurVisId, callback, changeIsVal
...
     return ENABLED_VIS_TYPES.map((vs: string) => {
-      if (isLineOrScatter(vs)) {
-        visDefinition = vs === visChartTypes.Line ? getVisType(vs, { type: visChartTypes.Line }) : getVisType(vs, { type: visChartTypes.Scatter });
-      } else if (isVerticalorHorizontalBar(vs)) {
-        visDefinition = vs === visChartTypes.Bar ? getVisType(vs, { type: visChartTypes.Bar }) : getVisType(vs, { type: visChartTypes.HorizontalBar });
-      } else {
-        visDefinition = getVisType(vs);
-      }
-
+      visDefinition = getVisTypeData(vs);
```
Comment (spattnaik) [inline dashboards-observability/public/components/event_analytics/explorer/visualizations/config_panel/config_panel.tsx]: This `visDefinition` is going to be updated for each of the entry in the `ENABLED_VIS_TYPES` array. So you are only going to get the last one. In addition, your map function is not returning anything. Take a look. If you need the last element only, there is no need for a loop.

**Review by spattnaik — CHANGES_REQUESTED** [review]

File: dashboards-observability/public/components/event_analytics/explorer/visualizations/config_panel/config_panel.tsx (hunk 4 above)
Comment (deepaknevdepsl) [reply dashboards-observability/public/components/event_analytics/explorer/visualizations/config_panel/config_panel.tsx]: Optimized above block of code as per your suggestions.

File: dashboards-observability/public/components/event_analytics/explorer/visualizations/config_panel/config_panel.tsx (hunk 5 above)
Comment (deepaknevdepsl) [reply dashboards-observability/public/components/event_analytics/explorer/visualizations/config_panel/config_panel.tsx]: Optimized above block of code as per your suggestions.

**Comment by abasatwar** [comment]
https://github.com/opensearch-project/observability/pull/924 raised for this functionality.

**Pull request closed without merging**